File: 250313.cpp
#include <iostream>
#include <vector>
#include <string>
#include <map>
#include <algorithm>
#include <numeric>
#include <Eigen/Dense>
#include <iomanip>
#include <cmath>

// Helper: Print Eigen matrix
void print_matrix(const Eigen::MatrixXd& mat, const std::string& name = "") {
    if (!name.empty()) std::cout << name << ":\n";
    std::cout << mat << "\n";
}

int main() {
    // Series
    std::vector<int> s = {1,2,3,4,5,6,7,8,9,10};
    std::cout << "Series: ";
    for (auto v : s) std::cout << v << " ";
    std::cout << std::endl;
    std::cout << "Head: ";
    for (int i = 0; i < 3; ++i) std::cout << s[i] << " ";
    std::cout << std::endl;
    std::cout << "Tail: ";
    for (int i = s.size()-3; i < s.size(); ++i) std::cout << s[i] << " ";
    std::cout << std::endl;

    // DataFrame
    Eigen::MatrixXd df = Eigen::MatrixXd::Random(6,4); // 6x4 random
    std::vector<int> df_index = {11,12,14,15,16,17};
    std::vector<std::string> df_columns = {"A","B","C","D"};
    print_matrix(df, "df");

    // Column selection
    std::cout << "df['A']: ";
    for (int i = 0; i < df.rows(); ++i) std::cout << df(i,0) << " ";
    std::cout << std::endl;
    std::cout << "df[['A','B']]:\n";
    for (int i = 0; i < df.rows(); ++i) std::cout << df(i,0) << ", " << df(i,1) << std::endl;

    // dtypes
    std::cout << "All columns are double (Eigen::MatrixXd)" << std::endl;

    // Tail
    std::cout << "df.tail():\n";
    for (int i = df.rows()-3; i < df.rows(); ++i) {
        for (int j = 0; j < df.cols(); ++j) std::cout << df(i,j) << " ";
        std::cout << std::endl;
    }

    // Index, columns, values
    std::cout << "Index: ";
    for (auto idx : df_index) std::cout << idx << " ";
    std::cout << std::endl;
    std::cout << "Columns: ";
    for (auto col : df_columns) std::cout << col << " ";
    std::cout << std::endl;
    std::cout << "Values:\n";
    print_matrix(df);

    // Describe
    std::cout << "Describe:\n";
    for (int j = 0; j < df.cols(); ++j) {
        double mean = df.col(j).mean();
        double stddev = std::sqrt((df.col(j).array() - mean).square().sum() / (df.rows()-1));
        double min = df.col(j).minCoeff();
        double max = df.col(j).maxCoeff();
        std::cout << df_columns[j] << ": mean=" << mean << ", std=" << stddev << ", min=" << min << ", max=" << max << std::endl;
    }

    // Transpose
    print_matrix(df.transpose(), "df.T");

    // Sort columns (descending)
    std::cout << "Sort columns descending:\n";
    Eigen::MatrixXd df_sorted = df;
    std::reverse(df_columns.begin(), df_columns.end());
    for (int j = 0; j < df.cols(); ++j) {
        for (int i = 0; i < df.rows(); ++i) {
            df_sorted(i,j) = df(i,df.cols()-1-j);
        }
    }
    print_matrix(df_sorted);

    // Sort by column B
    std::vector<std::pair<double,int>> b_vals;
    for (int i = 0; i < df.rows(); ++i) b_vals.push_back({df(i,1),i});
    std::sort(b_vals.begin(), b_vals.end());
    std::cout << "Sort by B:\n";
    for (auto& p : b_vals) {
        for (int j = 0; j < df.cols(); ++j) std::cout << df(p.second,j) << " ";
        std::cout << std::endl;
    }

    // Integer DataFrame
    Eigen::MatrixXi df2 = Eigen::MatrixXi::NullaryExpr(6,4,[](){return rand()%3;});
    std::vector<std::string> df2_index = {"a","b","c","d","e","f"};
    print_matrix(df2.cast<double>(), "df2");

    // Sorts
    // ...sorts omitted for brevity...

    // Logical operator
    std::cout << "df[df['A']>0]:\n";
    for (int i = 0; i < df.rows(); ++i) {
        if (df(i,0)>0) {
            for (int j = 0; j < df.cols(); ++j) std::cout << df(i,j) << " ";
            std::cout << std::endl;
        }
    }

    // Add new column
    std::vector<std::string> E = {"one","one","two","three","four","three"};
    std::vector<int> F(df2.rows());
    for (int i = 0; i < df2.rows(); ++i) F[i] = df2(i,0) + df2(i,1);
    std::cout << "E: "; for (auto& e : E) std::cout << e << " "; std::cout << std::endl;
    std::cout << "F: "; for (auto& f : F) std::cout << f << " "; std::cout << std::endl;

    // Statistics
    std::cout << "Mean per row:\n";
    for (int i = 0; i < df.rows(); ++i) {
        double mean = df.row(i).mean();
        std::cout << "Row " << i << ": " << mean << std::endl;
    }
    // ...other stats omitted for brevity...

    // Value counts (categorical)
    std::map<std::string,int> value_counts;
    for (auto& e : E) value_counts[e]++;
    std::cout << "Value counts for E:\n";
    for (auto& kv : value_counts) std::cout << kv.first << ": " << kv.second << std::endl;

    // Groupby
    std::map<std::string,int> group_sum;
    for (int i = 0; i < E.size(); ++i) group_sum[E[i]] += df2(i,0);
    std::cout << "Groupby E, sum of A:\n";
    for (auto& kv : group_sum) std::cout << kv.first << ": " << kv.second << std::endl;

    // Missing value handling, concat, merge, image section omitted for brevity
    return 0;
}

File: 250313.py
import pandas as pd # The prediction step provide correct image path
import numpy as np

s=pd.Series([1,2,3,4,5,6,7,8,9,10])
print(s)
print(s.head(3))
print(s.tail(3))
#visualize it in table 

np.random.randn(6,4)
df=pd.DataFrame(np.random.randn(6,4),index=[11,12,14,15,16,17],columns=['A','B','C','D'])
print(df)

#df[0]
#df[:,0]

print(df['A'])
print(df[['A','B']])
print(df.dtypes)

print(df.tail())
print(df.tail(3))

print(df.index)
print(df.columns)
print(df.values)
print(df.describe())
print(df.T)
print(df.sort_index(axis=1,ascending=False))
print(df.sort_values(by='B'))
print(df.to_numpy())
print(df.T)#index and column transpost
print(df.sort_index(axis=1,ascending=False))
print(df.sort_values(by='B'))
print(df.to_numpy())

df2=pd.DataFrame(np.random.randint(0,3,size=(6,4)),index=['a','b','c','d','e','f'],columns=['A','B','C','D']) #Lowest,highest, #,index=[11,12,14,15,16,17],columns=['A','B','C','D'])
print(df2.sort_values(by='A'))
print(df2.sort_values(by='B'))
print(df2.sort_values(by=['A','B']))
print(df2.sort_values(by=['A','B'],ascending=False))
print(df2.sort_values(by=['A','B'],ascending=[False,True]))
print(df.loc[[11,12]])
print(df.loc[[11,12],['A','B']])
print(df.iloc[[0,1]])
print(df.iloc[[0,1],[0,1]])
print(df.iloc[0:2,0:2])

#logical operator; every element that variable a is greater than 0
print(df[df['A']>0])
print(df2[df2['B'].isin([0,1])])#+ "is in the list")

df2['E']=['one','one','two','three','four','three'] #add new column
print(df2)
df2['F']=df2['A']+df2['B'] #Add new column by adding two columns which can be used in Feature Generation
print(df2)

print(type(df['A']))
s=pd.Series([1,2,3,4,5,6,7,8,9,10],index=[11,12,14,15,16,17,18,19,20,21])
print(s)
print(s[11])
df2['F']=s
print(s.values)
#df2['F']=s.values 
#print(df2)

df2['E']=[1,2,3,np.nan,5,6] #add new column
print(df2)
print(df2.dropna()) #drop row
print(df.fillna(value=0)) #fill missing value
print(df2['E'].fillna(value=df2['E'].mean())) #fill missing value
#drop column


#Unobserved value 


df.mean(axis=1) #If you want to calculate mean of individual row
df.var(axis=1)
df.std(axis=1)
df.sum(axis=1)
df.quantile(0.5,axis=1)#median
df.quantile(0.25,axis=1)#Q1
df.quantile(0.75,axis=1)#Q3
df.max(axis=1)
df.min(axis=1)
print(df.describe())



#For the categorial variable, tto check the distribution, check the frequency of the variable
df2['E'].value_counts()
df2['E'].value_counts(normalize=True)
df2['E'].value_counts(normalize=True)*100
print(df2['E'].value_counts(normalize=True)*100)


pd.concat((df,df2)) # Difference bewteen pd.concat((df,df2)) und pd.concat((df,df2),axis=1) Joining horizontally and vertically


left=pd.DataFrame({'key':['one','two','three','four'],'x1':np.random.rand(4),'x2':np.random.rand(4)})#certain keyvariable is used to join the two dataframes
right=pd.DataFrame({'key':['one','three','five','six','seven'],'x3':np.random.rand(4),'x4':np.random.rand(4)})

left.merge(right,on='key',how='inner') #inner join
left.merge(right,on='key',how='outer') #outer join
left.merge(right,on='key',how='left') #left join
left.merge(right,on='key',how='right') #right join


#Groupby
df2.groupby(['E'])['A'].sum()


""" Image produce und editing
from keras.preprocessing import image
test_image = image.load_img('image path', target_size=(64, 64))
test_image = image.img_to_array(test_image)
test_image = np.expand_dims(test_image, axis=0)
result = model.predict(test_image)
print(training_set.class_indices)
if result[0][0] == 1:
    prediction = 'Your class'"""